src/csharp/Microsoft.Spark/Sql/RowConstructor.cs

```diff
@@ -16,74 +18,116 @@ namespace Microsoft.Spark.Sql
     internal sealed class RowConstructor : IObjectConstructor
     {
         /// <summary>
-        /// Schema of the rows being received. Note that this is thread local variable
+        /// Cache the schema of the rows being received. Note that this is thread local variable
         /// because one RowConstructor object is registered to the Unpickler and there
         /// could be multiple threads unpickling the data using the same object registered.
         /// </summary>
         [ThreadStatic]
-        private static StructType s_currentSchema;
+        private static IDictionary<string, StructType> s_schemaCache;
 
         /// <summary>
-        /// Stores values passed from construct().
+        /// The RowConstructor that created this instance.
         /// </summary>
-        private object[] _values;
+        private readonly RowConstructor _parent;
 
         /// <summary>
-        /// Stores the schema for a row.
+        /// Stores the args passed from construct().
         /// </summary>
-        private StructType _schema;
+        private readonly object[] _args;
 
-        /// <summary>
-        /// Returns a string representation of this object.
-        /// </summary>
-        /// <returns>A string representation of this object</returns>
-        public override string ToString()
+        public RowConstructor() : this(null, null)
+        {
+        }
+
+        public RowConstructor(RowConstructor parent, object[] args)
         {
-            return string.Format("{{{0}}}", string.Join(",", _values));
+            _parent = parent;
+            _args = args;
         }
 
         /// <summary>
         /// Used by Unpickler to pass unpickled data for handling.
         /// </summary>
         /// <param name="args">Unpickled data</param>
-        /// <returns>New RowConstructor object capturing unpickled data</returns>
+        /// <returns>New RowConstructor object capturing args data</returns>
         public object construct(object[] args)
         {
             // Every first call to construct() contains the schema data. When
             // a new RowConstructor object is returned from this function,
             // construct() is called on the returned object with the actual
-            // row data.
+            // row data. The original RowConstructor object may be reused by the
+            // Unpickler and each subsequent construct() call can contain the
+            // schema data or a RowConstructor object that contains row data.
+            if (s_schemaCache is null)
+            {
+                s_schemaCache = new Dictionary<string, StructType>();
+            }
 
-            // Cache the schema to avoid parsing it for each construct() call.
-            if (s_currentSchema is null)
+            if ((args.Length == 1) && (args[0] is RowConstructor))
```

Comment:
can you do
`arg[0] is RowConstructor rowConstructor` and do `args[0] = rowConstructor.GetRow()`?

Reply:
Made suggested modifications.